11 fetch preprocess gdp

Juan-Mateos opened the pull request:
Closes #7 
Closes #12 
Closes #11

bishax committed:
Setup Nesta Data Science cookiecutter

bishax committed:
Do not create bucket on make init - already exists

Commit:
added a couple of background documents

bishax committed:
Add pragmatic flake8 config

bishax committed:
Phrasing change to nudge towards rebase

bishax committed:
fixup! Add pragmatic flake8 config

Commit:
added xlrd to requirements

Commit:
added scripts with sic preprocessing flow and utils

Commit:
added sic lookup getter

Commit:
added instructions to readme

bishax committed:
Make project default to 'project:industrial_taxonomy' metaflow namespace

Commit:
added script with utilities to fetch official data

Commit:
added scripts to fetch and process gdp data

Commit:
added the openpyxl requirement

Commit:
added type hints to gdp flow

Commit:
removed gdp specific method from excel_to_df function

Commit:
added dropna to process function

Commit:
added script to fetch and process population estimates

Commit:
utils to process nomis data

Commit:
added flow to collect nomis - aps and ashe - data

georgerichardson commented:
I got the following error log when I cloned the repo and ran `make install` from the `pip install xlrd` command

```
ERROR: Could not find a version that satisfies the requirement xldr (from industrial-taxonomy[dev]) (from versions: none)
ERROR: No matching distribution found for xldr
```

Commit:
Merge branch 'dev' into 11_fetch_preprocess_gdp

georgerichardson commented on industrial_taxonomy/pipeline/official/utils.py:
```diff
@@ -0,0 +1,21 @@
+    return response.content
+
+
+def excel_to_df(content: bytes, sheets: int, skiprows: int):
```
This throws `ImportError: Missing optional dependency 'openpyxl'.  Use pip or conda to install openpyxl` I think because [pandas now defaults to using openpyxl to open .xlsx files](https://github.com/pandas-dev/pandas/issues/38424). I think you either need to specify the engine or add openpyxl to `requirements.txt`

Separately, I'm not sure I see the value in this function as it's just a very thin wrapper around `pd.read_excel` that removes access to some parameters.

georgerichardson commented on README.md:
```diff
@@ -12,8 +12,6 @@
 
-### Pipeline
-
 Read instructions to reproduce the pipeline in `industrial_taxonomy/pipeline/README.md`.
```
Add equivalents for nomis, gdp and population?

georgerichardson commented on industrial_taxonomy/pipeline/official/gdp/flow.py:
```diff
@@ -0,0 +1,69 @@
+
+        # Create dfs for the sheets with relevant information (population and GVA)
+        self._pop_raw, self._gva_raw = [
+            excel_to_df(gdp_table, sheets=sh, skiprows=1) for sh in [7, 8]
```
More of a general question as to whether these kinds of hard coded variables should be kept in a config file?

georgerichardson commented on industrial_taxonomy/pipeline/official/nomis/flow.py:
```diff
@@ -0,0 +1,65 @@
+from industrial_taxonomy.pipeline.official.utils import get
+
+_APS_URL = "https://www.nomisweb.co.uk/api/v01/dataset/NM_17_5.data.csv?geography=1811939329...1811939332,1811939334...1811939336,1811939338...1811939497,1811939499...1811939501,1811939503,1811939505...1811939507,1811939509...1811939517,1811939519,1811939520,1811939524...1811939570,1811939575...1811939599,1811939601...1811939628,1811939630...1811939634,1811939636...1811939647,1811939649,181193965 ...
+_ASHE_URL = "https://www.nomisweb.co.uk/api/v01/dataset/NM_30_1.data.csv?geography=1811939329...1811939332,1811939334...1811939336,1811939338...1811939497,1811939499...1811939501,1811939503,1811939505...1811939507,1811939509...1811939517,1811939519,1811939520,1811939524...1811939570,1811939575...1811939599,1811939601...1811939628,1811939630...1811939634,1811939636...1811939647,1811939649,18119396 ...
```
Minor: could split into multiline

georgerichardson commented on industrial_taxonomy/pipeline/official/nomis/flow.py:
```diff
@@ -0,0 +1,65 @@
+@project(name="industrial_taxonomy")
+class NomisTables(FlowSpec):
+    """Flow to collect APS / ASHE data from NOMIS
+    Attributes:
```
```suggestion

    Attributes:
```

georgerichardson commented on industrial_taxonomy/pipeline/official/nomis/utils.py:
```diff
@@ -0,0 +1,37 @@
+    **kwargs,
+):
+    """Process nomis data
+    Arguments:
```
```suggestion

    Arguments:
```

georgerichardson commented on industrial_taxonomy/pipeline/official/gdp/utils.py:
```diff
@@ -0,0 +1,19 @@
+            axis=0, subset=["LA code"]  # We are dropping bottom rows without a LA code
+        )
+        .rename(columns={"2019\n[note 3]": "2019"})
+        .rename(columns={x: str(x) for x in range(1998, 2020)})
```
Perhaps something to address later: Will this be a common date range for datasets and later analysis?

georgerichardson commented on industrial_taxonomy/pipeline/official/nomis/utils.py:
```diff
@@ -0,0 +1,37 @@
+        value_column: value column
+        source: data source
+        indicator_column: column that contains the indicator
+    Returns:
```
```suggestion

    Returns:
```

georgerichardson commented on industrial_taxonomy/pipeline/official/population/flow.py:
```diff
@@ -0,0 +1,51 @@
+from metaflow import FlowSpec, project, step
+from metaflow.decorators import StepDecorator
+
+URL = (
```
Maybe a more specific name?

georgerichardson commented on industrial_taxonomy/pipeline/official/population/flow.py:
```diff
@@ -0,0 +1,51 @@
+from metaflow.decorators import StepDecorator
+
+URL = (
+    "https://www.ons.gov.uk/file?uri=/peoplepopulationandcommunity/populationandmigration/populationestimates/",
+    "datasets/populationestimatesforukenglandandwalesscotlandandnorthernireland/",
+    "mid2020/ukpopestimatesmid2020on2021geography.xls",
```
If the comma separation from these is removed then no need for `self.url = "".join(URL)` below

georgerichardson commented on industrial_taxonomy/pipeline/official/gdp/flow.py:
```diff
@@ -0,0 +1,69 @@
+from metaflow import FlowSpec, project, step
+
+URL = (
+    "https://www.ons.gov.uk/file?uri=/economy/grossdomesticproductgdp/datasets/",
+    "regionalgrossdomesticproductlocalauthorities/1998to2019/",
+    "regionalgrossdomesticproductlocalauthorities.xlsx",
```
If the comma separation from these is removed then no need for self.url = "".join(URL) below

georgerichardson commented on industrial_taxonomy/pipeline/official/gdp/utils.py:
```diff
@@ -0,0 +1,19 @@
+        .rename(columns={x: str(x) for x in range(1998, 2020)})
+    )
+
+    _table.columns = [re.sub(" ", "_", col.lower()) for col in _table.columns]
```
I think in general it's preferable to use Python string functions (in this case `.replace()`) over regex to avoid added complexity and for speed, although in this case it hardly makes a difference :D

georgerichardson reviewed (changes requested):
**General comments**
- Is there a need to have population data gathered from the GDP and population flows separately? Is this to get GPD per capita per year later?
- Minor: inconsitencies with whitespace between lines. Have noted a few
- Should we document the shapes of the data outputs from the flows either in the docstring or in the readme?

**What I did**
- [x] cloned the respository and ran `make install` on this branch from scratch
- [x] ran `python industrial_taxonomy/pipeline/sic/flow.py --production run`
- [x] ran `python industrial_taxonomy/pipeline/official/gdp/flow.py run` 
 - Did not finish. See xlrd error in `industrial_taxonomy/pipeline/official/utils.py` below
- [x] ran `python industrial_taxonomy/pipeline/official/nomis/flow.py run`
 - Finished running
 - Example output for `nomis_dict`
 ```
 {'date': 2019,
  'geography_name': 'Tower Hamlets',
  'geography_code': 'E09000030',
  'variable': 'Economic activity rate - aged 16-64',
  'value': 74.2,
  'source': 'aps'},
 {'date': 2019,
  'geography_name': 'Tower Hamlets',
  'geography_code': 'E09000030',
  'variable': 'Employment rate - aged 16-64',
  'value': 71.4,
  'source': 'aps'},
 {'date': 2019,
  'geography_name': 'Tower Hamlets',
  'geography_code': 'E09000030',
  'variable': '% with NVQ4+ - aged 16-64',
  'value': 60.5,
  'source': 'aps'},
 {'date': 2019,
  'geography_name': 'Tower Hamlets',
  'geography_code': 'E09000030',
  'variable': '% with other qualifications (NVQ) - aged 16-64',
  'value': 8.3,
  'source': 'aps'},
 {'date': 2019,
  'geography_name': 'Tower Hamlets',
  'geography_code': 'E09000030',
  'variable': '% with no qualifications (NVQ) - aged 16-64',
  'value': 9.7,
  'source': 'aps'},
  ```
- [x] ran `python industrial_taxonomy/pipeline/official/population/flow.py run`\
 - Finished running
 - Example output for `pop_est`
 ```
 ...
  'N09000006': 117337,
 'N09000007': 146452,
 'N09000008': 139443,
 'N09000009': 148953,
 'N09000010': 181669}
 ```

Commit:
added openpyxl to requirements

Commit:
Removed the excel_to_df function

Commit:
Replaced excel_to_df with pandas function

Commit:
Replaced excel_to_df from population flow

Commit:
added details to pipeline readme

Commit:
various fixes to flows

Juan-Mateos commented:
Thanks for the review! I have:

* Added `openpyxl` to requirements
* Added instructions to `README`
* Put hardcoded variables (excel sheets / year ranges) at the top of the script
* Made URL names more informative, split into multiline, removed commas
* Inserted whitespace before `Attributes`, `Arguments` and `Returns`
* Replaced regex with `.replace()`

Commit:
fixed bug in one of the variables being read

Commit:
added getters for official data

Juan-Mateos commented:
And then I added the `getters`.

bishax commented on industrial_taxonomy/__init__.py:
```diff
@@ -16,7 +16,6 @@ def get_yaml_config(file_path: Path) -> Optional[dict]:
     if file_path.exists():
         with open(file_path, "rt") as f:
             return yaml.load(f.read(), Loader=yaml.FullLoader)
-    return None
```
I think this somehow got removed when merging in `dev`

bishax replied:
Still need to add this back in

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+    try:
+        return next(filter(lambda run: run.successful, runs))
+    except StopIteration as exc:
+        raise MetaflowNotFound("Matching run not found") from exc
```
Note: #25 adds this as a general util and refactors existing uses - as this PR will probably merge first I can refactor in #25

https://github.com/nestauk/industrial_taxonomy/pull/25/files#diff-00b2db5bbabdce7be9d5602c61e3c8d66ac8628e428e6149234027c25230320d

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+# Data getters for official data
```
Don't forget to run `flake8`, it will flag up a lot of minor things to fix...

```bash
❯ flake8 industrial_taxonomy 
industrial_taxonomy/getters/official.py:4:1: F401 'typing.Dict' imported but unused
industrial_taxonomy/getters/official.py:4:1: F401 'typing.Optional' imported but unused
industrial_taxonomy/getters/official.py:6:1: F401 'metaflow.Run' imported but unused
industrial_taxonomy/getters/official.py:15:1: DAR201 Missing "Returns" in Docstring: - return
industrial_taxonomy/getters/official.py:15:1: DAR401 Missing exception(s) in Raises section: -r MetaflowNotFound
industrial_taxonomy/getters/official.py:27:1: DAR201 Missing "Returns" in Docstring: - return
industrial_taxonomy/getters/official.py:43:1: DAR201 Missing "Returns" in Docstring: - return
industrial_taxonomy/getters/official.py:59:1: DAR201 Missing "Returns" in Docstring: - return
industrial_taxonomy/getters/official.py:80:1: DAR201 Missing "Returns" in Docstring: - return
industrial_taxonomy/pipeline/official/utils.py:5:1: F401 'pandas as pd' imported but unused
industrial_taxonomy/pipeline/official/population/utils.py:2:1: F401 'pandas as df' imported but unused
industrial_taxonomy/pipeline/official/population/flow.py:4:1: F401 'metaflow.decorators.StepDecorator' imported but unused
industrial_taxonomy/pipeline/official/population/flow.py:7:112: B950 line too long (111 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/utils.py:14:1: DAR101 Missing parameter(s) in Docstring: - **kwargs
industrial_taxonomy/pipeline/official/nomis/flow.py:6:1: F401 'metaflow.decorators.StepDecorator' imported but unused
industrial_taxonomy/pipeline/official/nomis/flow.py:12:90: B950 line too long (89 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:13:101: B950 line too long (100 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:14:101: B950 line too long (100 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:15:102: B950 line too long (101 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:16:109: B950 line too long (108 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:17:113: B950 line too long (112 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:18:90: B950 line too long (89 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:22:90: B950 line too long (89 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:23:114: B950 line too long (113 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:24:112: B950 line too long (111 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:25:114: B950 line too long (113 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:26:110: B950 line too long (109 > 80 characters)
industrial_taxonomy/pipeline/official/nomis/flow.py:27:111: B950 line too long (110 > 80 characters)
industrial_taxonomy/pipeline/official/gdp/utils.py:4:1: F401 're' imported but unused
```

bishax replied:
For a top-level description use doc-string style rather than a `#` comment.
Anyone building API documentation with tools like Sphinx will then get a module-level docstring

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+
+@lru_cache()
+def get_run(flow: str):
+    """Last successful run executed with `--production`.
```
Here `flake8` raises the following:
```
industrial_taxonomy/getters/official.py:15:1: DAR201 Missing "Returns" in Docstring: - return
industrial_taxonomy/getters/official.py:15:1: DAR401 Missing exception(s) in Raises section: -r MetaflowNotFound
```

This might seem a bit anal but is useful to have this info for complex functions that may be used across modules.

For simpler functions with:
* Obvious function name
* A small number of obvious parameter names like this (suggest that `flow_name` is better than `flow` as it's less likely someone using it erroneously thinks they need to pass in a `metaflow.Flow` object) 
* Type annotations (suggest adding `-> Run` for the return type)

our flake8 config allows you to write a one-line docstring and not get nagged for Args/Returns/Raises

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+def gva_pc_lad():
+    """Get the GVA per capita in a local authority
+
+    Columns:
```
This and subsequent functions should be of the form

```
Returns:
  Columns:
     ...
```

so that API docs can be generated if needed and to make it clear that we're talking about a return dataframe

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+    )
+
+
+def population_lad():
```
I would recommend taking an optional `run` parameter like in the SIC getters...
 https://github.com/nestauk/industrial_taxonomy/blob/dev/industrial_taxonomy/getters/sic.py

This allows one to pass a non `--production` run to the getter which is great for:
- Iterative development - As you're iterating on a flow you can pass in a non `--production` run 
- Reviewers - Can run a flow in test mode and pass the resulting run into getters

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+    """get the population in a local authority
+
+    Columns:
+        Name: itl1_region, dtype: str, NUTS1 region (e.g. Scotland, South East etc)
```
`itl1_region` isn't a great name (I know you probably didn't choose it).

`nuts1_name` would be better because:
- It's more specific but not less accurate
- It makes it more clear that we're talking about the human readable name than the code name

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+    Columns:
+        Name: itl1_region, dtype: str, NUTS1 region (e.g. Scotland, South East etc)
+        Name: la_code, dtype: str, local authority code
+        Name: la_name, dtype: str, local authority name
```
General point:
It would be nice to have column naming guidelines across the practice. Both general naming guidelines but also a default for commonly used things like Local authority codes/names (different official data-sources go for `la_code`, `laua_code`, `LAD19_CD` and it drives me mad)

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+        Name: la_code, dtype: str, local authority code
+        Name: la_name, dtype: str, local authority name
+        Name: year dtype: int, year (ranges between 1998 and 2019
+        Name: gva dtype: float, GDP per capita
```
Are these columns right? `gdp_pc` is missing at least

bishax commented on industrial_taxonomy/pipeline/official/gdp/flow.py:
```diff
@@ -0,0 +1,71 @@
+"""Fetch GDP data"""
+
+from typing import Dict, List, Union
+import pandas as pd
```
Imports that aren't from the Python standard library or metaflow itself should go inside the function bodies of steps.

https://nestauk.github.io/kuebiko/guides/metaflow/#writing-flows
> Imports used within a step should always go within a step (at the top of it)
Because metaflow has the ability to abstract infrastructure then different steps may run on different machines with different environments. from kuebiko.pipeline.sic.utils import * wouldn't work in a step - kuebiko isn't installed/packaged in the new conda environment

bishax commented on industrial_taxonomy/pipeline/official/gdp/flow.py:
```diff
@@ -0,0 +1,71 @@
+    @step
+    def transform(self):
+        """Clean up the data"""
+        from utils import process_gdp_table
```
Import convention is inconsistent with L38

bishax commented on industrial_taxonomy/pipeline/official/utils.py:
```diff
@@ -0,0 +1,12 @@
+    """get ONS table in URL"""
+    response = requests.get(url)
+    response.raise_for_status()
+    return response.content
```
A nitpick's note on reusability...
by returning the content rather than the response object then anywhere else we want to make a GET request but use another response property, e.g. `response.json()`, we'll have to write a new function that repeats 2 out of 3 lines.

Other nitpicks:
- The docstring says it gets ONS tables but it's actually more general than that
- A more accurate name would be `get_content`

bishax commented on industrial_taxonomy/pipeline/official/gdp/utils.py:
```diff
@@ -0,0 +1,21 @@
+
+
+def process_gdp_table(table: pd.DataFrame):
+    """Cleans up the gdp table"""
```
The docstring doesn't tell us anything more than the function name
```suggestion
    """Drop rows without LA codes and tidy column names."""
```

bishax commented on industrial_taxonomy/pipeline/official/gdp/utils.py:
```diff
@@ -0,0 +1,21 @@
+
+    _table.columns = [col.lower().replace(" ", "_") for col in _table.columns]
+
+    return _table
```
pandas-fu tip: `rename` can take a function too...

```suggestion
    return (
        table.dropna(axis=0, subset=["LA code"])
        .rename(columns={"2019\n[note 3]": "2019"})
        .rename(columns={x: str(x) for x in YEAR_RANGE})
        .rename(columns=lambda s: s.lower().replace(" ", "_"))
    )```

bishax commented on industrial_taxonomy/pipeline/official/nomis/utils.py:
```diff
@@ -0,0 +1,39 @@
+    value_column: str,
+    source: str,
+    indicator_column: str = "MEASURES_NAME",
+    **kwargs,
```
kwargs isn't used
```suggestion
```

bishax commented on industrial_taxonomy/pipeline/official/nomis/utils.py:
```diff
@@ -0,0 +1,39 @@
+        .rename(columns=str.lower)
+        .assign(
+            date=lambda df: [  # We parse APS dates returned in format "y-m"
+                d if type(d) == int else int(d.split("-")[0]) for d in df["date"]
```
[Not requesting the change but] it may have been a better choice to separate things out a little more if the logic between processing APS and ASHE differs rather than having conditional logic embedded deep within the processing

bishax commented on industrial_taxonomy/pipeline/official/population/flow.py:
```diff
@@ -0,0 +1,53 @@
+    def transform(self):
+        """Transform the population table"""
+
+        from utils import clean_popest
```
Inconsistent from L28

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,109 @@
+    )
+
+
+def nomis():
```
Q: from a user perspective - would it be better to have a getter for each indicator? i.e. will they be used all together or not?

bishax reviewed (commented):
There's a lot of minor comments here but this is much easier to read, review, and use than on previous projects :smile: 

One difficulty I ran into was in running the getters...
it looks like the last time you ran `LocalGdpData` and `NomisTables` you ran them with `--production` and `--datastore local` which means that when it came for me to load the data, it failed because it was trying to load a local path from your machine.
Don't combine `--datastore local` and `--production` - I've just added an [issue to metaflow_extensions](https://github.com/nestauk/metaflow_extensions/issues/27) to explore whether it's possible to make metaflow raise a warning when this happens.
I've rerun the flows with `--production` and S3 datastore so that the getters work for everyone again.

Commit:
address AB review comments

Juan-Mateos commented:
I did the following:
- [x] Ran `flake8` and fixed problems
- [x] Simplified doc-string for `get_run` and changed argument name
- [x] Changed top-level descriptions to doc-string
- [x] Added `Returns` to getters
- [x] Added `run` parameter to getters
- [x] Renamed `itl1_region` as `nuts1_name`
- [x] Fixed doc-string about columns being returned from `gva_pc_lad`
- [x] Made imports consistent
- [x] Changed `get` function so it returns a response object instead of content
- [x] Made `process_gdp_table` more informative
- [x] Chained `rename` in `process_gdp_table`
- [x] Removed unused `kwargs` argument from `process_nomis`
- [x] Re-ran all flows without `datastore=local`
- [x] Tested all getters

I didn't:
* Separate aps from ashe processing in the `process_nomis` util
* Create separate getters for nomis variables (they are used together so this isn't necessary here)

Note that:
* I had to keep the `pandas` import outside of the flows. It was required for the typing hints at the beginning of the flows. When I tried to `import pandas as pd` when declaring the flow I got the following error: ` Cannot pickle dump artifact named "pd"`. I suppose I could:
  * Remove the `pd.DataFrame` typing hints
  * Declare the typing hints somewhere else in the script (in the last step?)

bishax commented on industrial_taxonomy/getters/official.py:
```diff
@@ -0,0 +1,120 @@
+        .assign(gva_pc=lambda df: (1e6 * df["gva"] / df["pop"]).round(2))
+        .rename(columns={"itl1_region": "nuts1_name"})
+        .drop(axis=1, labels=["gva", "pop"])
+    )
```
`itl1_region` is renamed to `nuts1_name` for only 1 out of 3 getters

bishax reviewed (commented):
For flake8 I get (I think you added the run arg after running flake8 so missed these)
>industrial_taxonomy/getters/official.py:23:1: DAR101 Missing parameter(s) in Docstring: - run
industrial_taxonomy/getters/official.py:44:1: DAR101 Missing parameter(s) in Docstring: - run
industrial_taxonomy/getters/official.py:89:1: DAR101 Missing parameter(s) in Docstring: - run

---

There's then 1 comment around the fact that there's a few `itl1_region`s still around.

---
On the `pandas` imports, I had to work around the same issue in #25 as follows (follow link)...
https://github.com/nestauk/industrial_taxonomy/blob/84cc4da7e5bab23bf8926596ae9461d91537eb3f/industrial_taxonomy/pipeline/glass_house/flow.py#L19
There is an import within flow steps as well as a top-level import within a try-except block that ignores `ImportError`s that may happen when executing the flow in a remote environment like Batch before it has installed pandas (it has to run the file to understand the structure of the flow and the dependencies it needs).
Flow-level type annotations are of the form so that static analysis tools like `flake8` and `mypy` can still resolve the imports but at runtime we don't get an error on Batch because the **object** `DataFrame` doesn't exist (because we got an `ImportError` above and so `DataFrame` isn't defined)
```python
class MyFlow(FlowSpec):
  some_data_artifact: "DataFrame"
```

---


After the above, then:
* Re-run the flows with `--production`
* Get latest changes from `dev`

Then all good from me.

Commit:
addressed AB review comments

Commit:
added return None

Juan-Mateos commented:
I did the following:
- [x] Fix new `flake8` problems
- [x] Include the typing import 'hack' in the `gdp` and `nomis` flows
- [x] Change `itl1_region` names in all getters.
- [x] Added `None` to `industrial_taxonomy/__init__.py`.
- [x] Re-ran all the flows
- [x] Tested the getters
